be/src/formats/parquet/scalar_column_reader.cpp

```diff
@@ -426,4 +429,181 @@ Status ScalarColumnReader::fill_dst_column(ColumnPtr& dst, ColumnPtr& src) {
+    DCHECK(get_column_parquet_field()->is_nullable ? dst->is_nullable() : true);
+    ColumnContentType content_type = ColumnContentType::DICT_CODE;
+
+    if (_dict_code == nullptr) {
+        _dict_code = ColumnHelper::create_column(
+                TypeDescriptor::from_logical_type(ColumnDictFilterContext::kDictCodePrimitiveType), true);
+    }
+    _ori_column = dst;
+    dst = _dict_code;
+    dst->reserve(range.span_size());
+
+    {
+        SCOPED_RAW_TIMER(&_opts.stats->column_read_ns);
+        return _reader->read_range(range, filter, content_type, dst.get());
+    }
+}
+
+bool LowCardColumnReader::try_to_use_dict_filter(ExprContext* ctx, bool is_decode_needed, const SlotId slotId,
+                                                 const std::vector<std::string>& sub_field_path, const size_t& layer) {
+    if (sub_field_path.size() != layer) {
+        return false;
+    }
+
+    if (column_all_pages_dict_encoded()) {
+        if (_dict_filter_ctx == nullptr) {
+            _dict_filter_ctx = std::make_unique<ColumnDictFilterContext>();
+            _dict_filter_ctx->is_decode_needed = is_decode_needed;
+            _dict_filter_ctx->sub_field_path = sub_field_path;
+            _dict_filter_ctx->slot_id = slotId;
+        }
+        _dict_filter_ctx->conjunct_ctxs.push_back(ctx);
+        return true;
+    } else {
+        return false;
+    }
+}
+
+Status LowCardColumnReader::fill_dst_column(ColumnPtr& dst, ColumnPtr& src) {
+    if (!_code_convert_map.has_value()) {
+        RETURN_IF_ERROR(_check_current_dict());
+    }
+
+    dst->resize(src->size());
+
+    const ColumnPtr& dict_codes = src;
+    auto* codes_nullable_column = ColumnHelper::as_raw_column<NullableColumn>(dict_codes);
+    auto* codes_column = ColumnHelper::as_raw_column<FixedLengthColumn<int32_t>>(codes_nullable_column->data_column());
+    const NullData& null_data_ptr = codes_nullable_column->immutable_null_column_data();
+
+    auto& codes = codes_column->get_data();
+    if (codes_nullable_column->has_null()) {
+        for (size_t i = 0; i < src->size(); i++) {
+            // if null, we assign dict code 0
+            // null = 0, mask = 0xffffffff
+            // null = 1, mask = 0x00000000
+            uint32_t mask = ~(static_cast<uint32_t>(-null_data_ptr[i]));
+            codes[i] = mask & codes[i];
+        }
+    }
+
+    auto* dst_data_column = down_cast<LowCardDictColumn*>(ColumnHelper::get_data_column(dst.get()));
+    SIMDGather::gather(dst_data_column->get_data().data(), _code_convert_map->data(), codes.data(),
+                       DICT_DECODE_MAX_SIZE, src->size());
+
+    if (dst->is_nullable()) {
+        auto* nullable_codes = down_cast<NullableColumn*>(dict_codes.get());
+        auto* nullable_dst = down_cast<NullableColumn*>(dst.get());
+        nullable_dst->null_column_data().swap(nullable_codes->null_column_data());
+        nullable_dst->set_has_null(nullable_codes->has_null());
+    }
+
+    src->reset_column();
```

Comment:
Why reset_column?

Reply:
src is used as temp column to store intermediate result, fill_dst_column means we have got a chunk of data, next time we call get_next, we will append data to it.